contracts/AssetPool.sol

```diff
@@ -80,17 +124,112 @@ contract AssetPool is Ownable {
+    ///         completed the withdrawal on behalf of the underwriter.
+    function completeWithdrawal(address underwriter) external {
+        /* solhint-disable not-rely-on-time */
+        uint256 initiatedAt = withdrawalInitiatedTimestamp[underwriter];
+        require(initiatedAt > 0, "No withdrawal initiated for the underwriter");
+
+        uint256 withdrawalDelayEndTimestamp = initiatedAt.add(withdrawalDelay);
+        require(
+            withdrawalDelayEndTimestamp < block.timestamp,
+            "Withdrawal delay has not elapsed"
+        );
+
+        // slither-disable-next-line reentrancy-events
         rewardsPool.withdraw();
 
+        uint256 covAmount = pendingWithdrawal[underwriter];
         uint256 covSupply = underwriterToken.totalSupply();
         uint256 collateralBalance = collateralToken.balanceOf(address(this));
 
         uint256 amountToWithdraw =
             covAmount.mul(collateralBalance).div(covSupply);
 
-        underwriterToken.safeTransferFrom(msg.sender, address(this), covAmount);
+        //
+        //      withdrawal              graceful withdrawal
+        //        delay                 timeout
+        //  /--------------\ /----------|
+        // x----------------x-----------x-------------------------------x------>
+        // ^                 \-----------------------------------------|
+        // initiatedAt                                   hard withdrawal
+        //                                               timeout
+        //
+
+        // When the graceful withdrawal time ends. After this time, part of the
+        // collateral and rewards will be seized by the pool.
+        uint256 gracefulWithdrawalEndTimestamp =
+            withdrawalDelayEndTimestamp.add(gracefulWitdrawalTimeout);
+        // When the time for the withdrawal ends. After this time, 99% of
+        // rewards and collateral is seized by the pool and 1% of rewards and
+        // collateral is sent to the notifier who completed the withdrawal on
+        // behalf of the underwriter.
+        uint256 hardWithdrawalEndTimestamp =
+            withdrawalDelayEndTimestamp.add(hardWithdrawalTimeout);
+
+        if (gracefulWithdrawalEndTimestamp >= block.timestamp) {
+            // Before the graceful withdrawal timeout. This is the happy path.
+            emit WithdrawalCompleted(
+                underwriter,
+                amountToWithdraw,
+                block.timestamp
+            );
+            collateralToken.safeTransfer(underwriter, amountToWithdraw);
+        } else if (hardWithdrawalEndTimestamp >= block.timestamp) {
+            // After the graceful withdrawal timeout but before the hard
+            // withdrawal timeout. A portion of collateral and tokens is
+            // seized by the pool, proportionally to the time passed after
+            // the graceful withdrawal timeout.
+            uint256 delayRatio =
+                hardWithdrawalEndTimestamp.sub(block.timestamp).mul(1e18).div(
```

Comment:
unlikely but we can have a corner case when `hardWithdrawalEndTimestamp == block.timestamp` and delayRatio = 0. Other transfers and calculations except emitting events won't make sense.

Reply:
Ooof, good catch! Fixed.